ui/utils.py

```diff
@@ -65,27 +82,65 @@ def get_moira_user(user):
     return MoiraUser(user.email, 'STRING')
 
 
-def user_moira_lists(user):
+def set_moira_cache(user, membership):
+    """
+    Set the user's moira lists cache
+
+    Args:
+        user (django.contrib.auth.User): the Django user to return a Moira user for.
+        membership (UserMoiraMembership): object specifying which lists a member is or isn't on.
+    """
+    cache.set(MOIRA_CACHE_KEY.format(user_id=user.id), membership, settings.MOIRA_CACHE_TIMEOUT)
+
+
+def delete_moira_cache(user):
+    """
+    Delete the user's moira list cache
+
+    Args:
+        user (django.contrib.auth.User): the Django user to return a Moira user for.
+    """
+    cache.delete(MOIRA_CACHE_KEY.format(user_id=user.id))
+
+
+def query_moira_lists(user):
     """
-    Get a list of all the moira lists a user has access to.
+    Get a UserMoiraMembership object with all the moira lists a user has access to, by querying the Moira service.
 
     Args:
         user (django.contrib.auth.User): the Django user.
 
     Returns:
-        list: A list of moira lists the user has access to.
+        UserMoiraMembership: A list of moira lists the user has access to.
     """
     moira_user = get_moira_user(user)
     moira = get_moira_client()
     try:
-        return moira.user_lists(moira_user.username, moira_user.type)
+        membership = UserMoiraMembership(member_of=moira.user_lists(moira_user.username, moira_user.type))
+        set_moira_cache(user, membership)
+        return membership
     except Exception as exc:  # pylint: disable=broad-except
         if 'java.lang.NullPointerException' in str(exc):
             # User is not a member of any moira lists, so ignore exception and return empty list
-            return []
+            membership = UserMoiraMembership()
+            set_moira_cache(user, membership)
+            return membership
         raise MoiraException('Something went wrong with getting moira-lists for %s' % user.username) from exc
 
 
+def user_moira_lists(user):
+    """
+    Get a list of all the moira lists a user has access to, from the cache if it exists, otherwise query Moira service.
+
+    Args:
+        user (django.contrib.auth.User): the Django user.
+
+    Returns:
+        UserMoiraMembership: An object containing all known lists the user does & doesn't have access to.
+    """
+    return cache.get(MOIRA_CACHE_KEY.format(user_id=user.id)) or query_moira_lists(user)
+
+
 def has_common_lists(user, list_names):
```

Comment:
This function now has a side effect of altering the moira_cache. It may make sense to note this in the function's docstring. Your call.